p2p/cuprate-p2p/src/block_downloader.rs

```diff
@@ -0,0 +1,1181 @@
+
+        for client in self.client_pool.borrow_clients(&peers) {
+            self.pending_peers
+                .entry(client.info.pruning_seed)
+                .or_default()
+                .push(client);
+        }
+
+        self.check_pending_peers(chain_tracker).await;
+
+        Ok(())
+    }
+
+    /// Checks if we have batches ready to send down the [`BufferAppender`].
+    ///
+    /// We guarantee that blocks sent down the buffer are sent in the correct order.
+    async fn push_new_blocks(&mut self) -> Result<(), BlockDownloadError> {
+        while let Some(ready_batch) = self.ready_batches.peek() {
+            // Check if this ready batch's start height is higher than the lowest in flight request.
+            // If there is a lower start height in the inflight requests then this is _not_ the next batch
+            // to send down the buffer.
+            if self
+                .inflight_requests
+                .first_key_value()
+                .is_some_and(|(&lowest_start_height, _)| {
+                    ready_batch.start_height > lowest_start_height
+                })
+            {
+                break;
+            }
+
+            // Our next ready batch is older (lower height) than the oldest in flight, push it down the
+            // buffer.
+            let ready_batch = self.ready_batches.pop().unwrap();
+
+            let size = ready_batch.block_batch.size;
+            self.ready_batches_size -= size;
+
+            tracing::debug!(
+                "Pushing batch to buffer, new ready batches size: {}",
+                self.ready_batches_size
+            );
+
+            self.buffer_appender
+                .send(ready_batch.block_batch, size)
+                .await
+                .map_err(|_| BlockDownloadError::BufferWasClosed)?;
+
+            // Loops back to check the next oldest ready batch.
+        }
+
+        Ok(())
+    }
+
+    /// Handles a response to a request to get blocks from a peer.
+    async fn handle_download_batch_res(
+        &mut self,
+        start_height: u64,
+        res: Result<(ClientPoolDropGuard<N>, BlockBatch), BlockDownloadError>,
+        chain_tracker: &mut ChainTracker<N>,
+    ) -> Result<(), BlockDownloadError> {
+        tracing::debug!("Handling block download response");
+
+        match res {
+            Err(e) => {
+                if matches!(e, BlockDownloadError::ChainInvalid) {
+                    // If the chain was invalid ban the peer who told us about it and error here to stop the
+                    // block downloader.
+                    self.inflight_requests.get(&start_height).inspect(|entry| {
+                        tracing::warn!(
+                            "Received an invalid chain from peer: {}, exiting block downloader (it should be restarted).",
+                            entry.peer_who_told_us
+                        );
+                        entry.peer_who_told_us_handle.ban_peer(LONG_BAN)
```

Comment:
nit but only kinda: https://stackoverflow.com/questions/26665471/are-semicolons-optional-in-rust.

`;` shows intent, i.e. "The returning value of this expression is ()". If `ban_peer()` eventually returned something other than `()` then this line would have different semantics, which wasn't the original intent.
```suggestion
                        entry.peer_who_told_us_handle.ban_peer(LONG_BAN);
```